File: flutter_app/lib/presentation/log/care_log_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:uuid/uuid.dart';

import '../../domain/model/data_models.dart';

class CareLogViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  List<DailyCareLog> _logs = [];
  List<DailyCareLog> get logs => _logs;

  bool _isLoading = false;
  bool get isLoading => _isLoading;

  void loadLogsForBooking(String bookingId) {
    _isLoading = true;
    notifyListeners();

    _firestore.collection("care_logs")
        .where("bookingId", isEqualTo: bookingId)
        .snapshots()
        .listen((snapshot) {
      final fetchedLogs = snapshot.docs.map((doc) => DailyCareLog.fromMap(doc.data())).toList();
      fetchedLogs.sort((a, b) => b.timestamp.compareTo(a.timestamp));
      
      _logs = fetchedLogs;
      _isLoading = false;
      notifyListeners();
    }, onError: (_) {
      _isLoading = false;
      notifyListeners();
    });
  }

  Future<void> addLog({
    required String bookingId,
    required String elderId,
    required String bloodPressure,
    required String sugarLevel,
    required String temperature,
    required String mealStatus,
    required bool medicationGiven,
    required String notes,
    required Function(bool, String) onResult,
  }) async {
    try {
      final caregiverId = _auth.currentUser?.uid;
      if (caregiverId == null) return;

      final logId = const Uuid().v4();
      final log = DailyCareLog(
        id: logId,
        bookingId: bookingId,
        caregiverId: caregiverId,
        elderId: elderId,
        timestamp: DateTime.now().millisecondsSinceEpoch,
        bloodPressure: bloodPressure,
        sugarLevel: sugarLevel,
        temperature: temperature,
        mealStatus: mealStatus,
        medicationGiven: medicationGiven,
        notes: notes,
      );

      await _firestore.collection("care_logs").doc(logId).set(log.toMap());
      onResult(true, "Log added successfully");
    } catch (e) {
      onResult(false, e.toString());
    }
  }

  // NOTE: PDF generation logic using `pdf` package would be triggered from UI and consume this data.
}

File: app/src/main/java/com/kina/care/presentation/log/CareLogViewModel.kt
package com.kina.care.presentation.log

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.kina.care.domain.model.DailyCareLog
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

class CareLogViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    private val _logs = MutableStateFlow<List<DailyCareLog>>(emptyList())
    val logs: StateFlow<List<DailyCareLog>> = _logs.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    // Load logs for a specific booking
    fun loadLogsForBooking(bookingId: String) {
        _isLoading.value = true
        firestore.collection("care_logs")
            .whereEqualTo("bookingId", bookingId)
            .orderBy("timestamp", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, e ->
                _isLoading.value = false
                if (e != null || snapshot == null) return@addSnapshotListener
                val fetchedLogs = snapshot.documents.mapNotNull { it.toObject(DailyCareLog::class.java)?.copy(id = it.id) }
                _logs.value = fetchedLogs
            }
    }

    fun addLog(
        bookingId: String,
        elderId: String,
        bloodPressure: String,
        sugarLevel: String,
        temperature: String,
        mealStatus: String,
        medicationGiven: Boolean,
        notes: String,
        onResult: (Boolean, String) -> Unit
    ) {
        val caregiverId = auth.currentUser?.uid ?: return
        val logId = firestore.collection("care_logs").document().id
        
        val log = DailyCareLog(
            id = logId,
            bookingId = bookingId,
            caregiverId = caregiverId,
            elderId = elderId,
            timestamp = System.currentTimeMillis(),
            bloodPressure = bloodPressure,
            sugarLevel = sugarLevel,
            temperature = temperature,
            mealStatus = mealStatus,
            medicationGiven = medicationGiven,
            notes = notes
        )

        firestore.collection("care_logs").document(logId).set(log)
            .addOnSuccessListener { onResult(true, "Log added successfully") }
            .addOnFailureListener { e -> onResult(false, e.message ?: "Failed to add log") }
    }
}
